File: C++/题目1518：反转链表.cpp

// 题目1518：反转链表

/**
 * @author:wangzq
 * @email:wangzhenqing1008@163.com
 * @date:2015-06-30 11:01:55
 * @url:http://ac.jobdu.com/problem.php?pid=1518
 */

import java.io.BufferedReader;
#include <stdio.h>
#include <malloc.h>

struct LinkedNode {
    int data;
    struct LinkedNode *next;
};

int n, i;

int main() {
    LinkedNode *node, *tail, *p, *q;
    while (scanf("%d", &n) != EOF) {
        if (n == 0) {
            printf("NULL\n");
            continue;
        }
        node = (LinkedNode *) malloc(sizeof(LinkedNode));
        node->next = NULL;
        tail = node;
        for (i = 0; i < n; i++) {
            int data;
            scanf("%d", &data);
            p = (LinkedNode *) malloc(sizeof(LinkedNode));
            p->data = data;
            tail->next = p;
            tail = p;
        }
        tail->next = NULL;
        p = node->next;
        tail = NULL;
        while (p != NULL) {
            q = p->next;
            p->next = tail;
            tail = p;
            p = q;
        }
        p = tail;
        for (i = 0; i < n; i++) {
            if (i == n - 1) {
                printf("%d\n", p->data);
            } else {
                printf("%d ", p->data);
                p = p->next;
            }
        }
    }
    return 0;
}
/**************************************************************
Problem: 1518
User: wangzhenqing
Language: C++
Result: Accepted
Time:150 ms
Memory:2472 kb
****************************************************************/


                        

File: Java/题目1518：反转链表.java

// 题目1518：反转链表

/**
 * @author:wangzq
 * @email:wangzhenqing1008@163.com
 * @date:2015-06-30 11:01:55
 * @url:http://ac.jobdu.com/problem.php?pid=1518
 */

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.StreamTokenizer;

public class Main {
    /*
     * 1518
     */
    public static void main(String[] args) throws Exception {
        StreamTokenizer st = new StreamTokenizer(new BufferedReader(
                new InputStreamReader(System.in)));
        while (st.nextToken() != StreamTokenizer.TT_EOF) {
            int n = (int) st.nval;
            if (n == 0) {
                System.out.println("NULL");
                continue;
            }
            LinkedNode node = null;
            LinkedNode point = null;
            for (int i = 0; i < n; i++) {
                st.nextToken();
                LinkedNode tempNode = new LinkedNode((int) st.nval, null);
                if (point == null) {
                    node = tempNode;
                    point = node;
                } else {
                    point.next = tempNode;
                    point = point.next;
                }
            }

            LinkedNode end = null;
            LinkedNode q;
            point = node;
            while (point != null) {
                q = point.next;
                point.next = end;
                end = point;
                point = q;
            }
            StringBuffer sb = new StringBuffer();
            while (end != null) {
                sb.append(end.data + " ");
                end = end.next;
            }
            System.out.println(sb.toString().trim());
        }
    }

    static class LinkedNode {
        private int data;

        private LinkedNode next;

        public LinkedNode(int data, LinkedNode next) {
            super();
            this.data = data;
            this.next = next;
        }

        public LinkedNode() {
            super();
        }
    }
}
/**************************************************************
 Problem: 1518
 User: wangzhenqing
 Language: Java
 Result: Accepted
 Time:1620 ms
 Memory:36964 kb
 ****************************************************************/




